frame/treasury/src/lib.rs

```diff
@@ -531,9 +701,94 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		r
 	}
 
-	/// Spend some money! returns number of approvals before spend.
+	/// Spend_funds is triggered periodically and uses the `T::Paymaster` to payout all spend
+	/// requests in the `PendingPayments` storage map.
 	pub fn spend_funds() -> Weight {
 		let mut total_weight = Weight::zero();
+		let mut total_spent = BalanceOf::<T, I>::zero();
+		let mut missed_proposals: u32 = 0;
+		let proposals_len = PendingPayments::<T, I>::count();
+
+		Self::deposit_event(Event::ProcessingProposals { waiting_proposals: proposals_len });
+
+		for key in PendingPayments::<T, I>::iter_keys() {
+			if let Some(mut p) = PendingPayments::<T, I>::get(key) {
+				match p.payment_id {
+					None => match T::Paymaster::pay(&p.beneficiary, p.asset_kind, p.value) {
+						Ok(id) => {
+							total_spent += p.normalized_value;
+							p.payment_id = Some(id);
+							Self::deposit_event(Event::PaymentTriggered {
+								pending_payment_index: key,
+								asset_kind: p.asset_kind,
+								amount: p.value,
+								payment_id: id,
+							});
+							PendingPayments::<T, I>::set(key, Some(p));
+						},
+						Err(err) => {
+							log::debug!(target: LOG_TARGET, "Paymaster::pay failed for PendingPayment with index: {:?} and error: {:?}", key, err);
+							missed_proposals = missed_proposals.saturating_add(1);
+							Self::deposit_event(Event::PaymentFailure {
+								pending_payment_index: key,
+								asset_kind: p.asset_kind,
+								amount: p.value,
+								payment_id: None,
+							});
+						},
+					},
+					Some(payment_id) => match T::Paymaster::check_payment(payment_id) {
```

Comment:
pallet_xcm does not timeout now. but I think it should be on both sides

Reply:
what do you propose?
With the current implementation, we would simply keep retrying the transaction each spend period until it succeeds, and that seems sufficient to me. 

Also, the treasury pallet knows nothing about the XCM timeouts nor does it even know anything about XCM. 
Eg lots of parachains will implement their treasuries using the `PayFromAccount` or `PayFungibles` implementations which have no timeouts. So i'm not convinced about adding any timeout logic into the substrate treasury implementation, as that is too specific to XCM

Reply:
I added a concept of retries, so we retry a payment a given number of times. 
About the timeouts, It would only be useful in the XCM context, and in the case of XCM, when it expires, the `check_payment` method would return a status unknown, which we now treat as a failure, so it would fail immediately at that point. I think with that workflow, there won't be any extra utility for the timeouts, and if that use-case comes up, then we can include it in a subsequent PR.